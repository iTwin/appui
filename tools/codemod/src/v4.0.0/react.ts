/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
import type { API, FileInfo, Options } from "jscodeshift";
import { useExtensions } from "../utils/Extensions";

export default function transformer(
  file: FileInfo,
  api: API,
  options: Options
) {
  const j = api.jscodeshift;
  useExtensions(j);

  const root = j(file.source);

  root.rename(
    "@itwin/appui-react:CommonWidgetProps",
    "@itwin/appui-react:Widget"
  );
  root.rename("@itwin/appui-react:WidgetProps", "@itwin/appui-react:Widget");

  return root.toSource(options.printOptions);
}
