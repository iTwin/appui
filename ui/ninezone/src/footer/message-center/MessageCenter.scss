/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/ui-core/colors";
@import "../_variables";

.nz-footer-messageCenter-messageCenter {
  min-width: $nz-footer-popover-min-width;
  transform: translateX(-50%);
  color: $background-5;
}
