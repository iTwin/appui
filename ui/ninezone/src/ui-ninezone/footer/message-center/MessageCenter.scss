/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
@import "@bentley/ui-core/lib/ui-core/style/themecolors";
@import "@bentley/ui-core/lib/ui-core/_z-index";
@import "../_variables";

.nz-footer-messageCenter-messageCenter {
  min-width: 400px;
  transform: translateX(-50%);
  color: $buic-background-control-stroke;

  @include uicore-z-index(status-popover);
}
