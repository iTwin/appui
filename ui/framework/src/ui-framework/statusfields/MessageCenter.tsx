/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Notification */

import * as React from "react";

import { NotifyMessageDetails, OutputMessagePriority } from "@bentley/imodeljs-frontend";
import { UiCore } from "@bentley/ui-core";

import { UiFramework } from "../UiFramework";

import { StatusBarFieldId } from "../widgets/StatusBarWidgetControl";
import { MessageManager } from "../messages/MessageManager";
import {
  MessageCenter, MessageCenterTab, MessageCenterMessage, MessageCenterDialog,
  TitleBarButton, FooterPopup,
} from "@bentley/ui-ninezone";
import { StatusFieldProps } from "./StatusFieldProps";

/** Enum for the [[MessageCenterField]] active tab
 * @internal
 */
enum MessageCenterActiveTab {
  AllMessages,
  Problems,
}

/** State for the [[MessageCenterField]] React component
 * @internal
 */
interface MessageCenterState {
  activeTab: MessageCenterActiveTab;
}

/** Properties of [[MessageCenterField]] component.
 * @public
 */
export interface MessageCenterFieldProps extends StatusFieldProps {
  /** Message center dialog target. */
  targetRef?: React.Ref<HTMLElement>;
}

/** Message Center Field React component.
 * @public
Â */
export class MessageCenterField extends React.Component<MessageCenterFieldProps, MessageCenterState> {
  private _className: string;
  private _target: React.MutableRefObject<HTMLDivElement | null> = {
    current: null,
  };
  private _indicator = React.createRef<HTMLDivElement>();

  public readonly state: Readonly<MessageCenterState> = {
    activeTab: MessageCenterActiveTab.AllMessages,
  };

  constructor(p: MessageCenterFieldProps) {
    super(p);

    const instance = this.constructor;
    this._className = instance.name;
  }

  public render(): React.ReactNode {
    const messageCount = MessageManager.messages.length;
    const footerMessages = (
      <>
        <MessageCenter
          className={this.props.className}
          style={this.props.style}
          indicatorRef={this._indicator}
          isInFooterMode={this.props.isInFooterMode}
          label={this.props.isInFooterMode ? UiFramework.translate("messageCenter.messages") : undefined}
          onClick={this._handleMessageIndicatorClick}
          targetRef={this._handleTargetRef}
        >
          {messageCount.toString()}
        </MessageCenter>
        <FooterPopup
          isOpen={this.props.openWidget === this._className}
          onClose={this._handleClose}
          onOutsideClick={this._handleOutsideClick}
          target={this._target}
        >
          <MessageCenterDialog
            buttons={
              <>
                <TitleBarButton title={UiFramework.translate("messageCenter.export")}>
                  <i className={"icon icon-export"} />
                </TitleBarButton>
                <TitleBarButton onClick={this._handleCloseMessageIndicatorClick} title={UiCore.translate("dialog.close")}>
                  <i className={"icon icon-close"} />
                </TitleBarButton>
              </>
            }
            prompt={UiFramework.translate("messageCenter.prompt")}
            tabs={
              <>
                <MessageCenterTab
                  isActive={this.state.activeTab === MessageCenterActiveTab.AllMessages}
                  onClick={() => this._changeActiveTab(MessageCenterActiveTab.AllMessages)}
                >
                  {UiFramework.translate("messageCenter.all")}
                </MessageCenterTab>
                <MessageCenterTab
                  isActive={this.state.activeTab === MessageCenterActiveTab.Problems}
                  onClick={() => this._changeActiveTab(MessageCenterActiveTab.Problems)}
                >
                  {UiFramework.translate("messageCenter.errors")}
                </MessageCenterTab>
              </>
            }
            title={UiFramework.translate("messageCenter.messages")}
          >
            {this.getMessages()}
          </MessageCenterDialog>
        </FooterPopup>
      </>
    );

    return footerMessages;
  }

  private _handleTargetRef = (instance: HTMLDivElement | null) => {
    if (typeof this.props.targetRef === "function")
      this.props.targetRef(instance);
    else if (this.props.targetRef)
      (this.props.targetRef as React.MutableRefObject<HTMLElement | null>).current = instance;
    this._target.current = instance;
  }

  private _handleClose = () => {
    this.setOpenWidget(null);
  }

  private _handleOutsideClick = (e: MouseEvent) => {
    if (!this._indicator.current ||
      !(e.target instanceof Node) ||
      this._indicator.current.contains(e.target))
      return;

    this._handleClose();
  }

  private _handleMessageIndicatorClick = () => {
    const isOpen = this.props.openWidget === this._className;
    if (isOpen)
      this.setOpenWidget(null);
    else
      this.setOpenWidget(this._className);
  }

  private _handleCloseMessageIndicatorClick = () => {
    this.setOpenWidget(null);
  }

  private _changeActiveTab = (tab: MessageCenterActiveTab) => {
    this.setState((_prevState) => {
      return {
        activeTab: tab,
      };
    });
  }

  private getMessages(): React.ReactChild[] {
    const messages = MessageManager.messages.slice(0).reverse();
    const tabRows: React.ReactChild[] = new Array<React.ReactChild>();

    messages.forEach((details: NotifyMessageDetails, index: number) => {
      /* istanbul ignore else */
      if (this.state.activeTab === MessageCenterActiveTab.AllMessages || this.isProblemStatus(details.priority)) {

        const iconClassName = MessageManager.getIconClassName(details);
        const message = details.briefMessage;

        tabRows.push(
          <MessageCenterMessage
            key={index.toString()}
            icon={<i className={iconClassName} />}
          >
            <span dangerouslySetInnerHTML={{ __html: message }} />
            {details.detailedMessage &&
              <>
                <br />
                <span className="uicore-text-small" dangerouslySetInnerHTML={{ __html: details.detailedMessage }} />
              </>
            }
          </MessageCenterMessage>,
        );
      }
    });

    return tabRows;
  }

  private isProblemStatus(priority: OutputMessagePriority): boolean {
    // See priority values in DgnPlatform defined in NotificationManager

    if (priority === OutputMessagePriority.Error || priority === OutputMessagePriority.Fatal)
      return true;

    return false;
  }

  private setOpenWidget(openWidget: StatusBarFieldId) {
    this.props.onOpenWidget(openWidget);
  }
}
