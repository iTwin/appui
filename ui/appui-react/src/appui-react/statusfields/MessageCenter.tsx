/*---------------------------------------------------------------------------------------------
 * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Notification
 */

import { OutputMessagePriority } from "@itwin/core-frontend";
import { Tab, Tabs } from "@itwin/itwinui-react";
import { Orientation } from "@itwin/core-react";
import type { CommonProps } from "@itwin/itwinui-react/cjs/core/utils";
import { Icon } from "@itwin/core-react";
import * as React from "react";
import { UiFramework } from "../UiFramework";
import { MessageCenterDialog } from "../layout/footer/message-center/Dialog";
import { MessageCenter } from "../layout/footer/message-center/Indicator";
import { MessageCenterMessage } from "../layout/footer/message-center/Message";
import { MessageCenterTab } from "../layout/footer/message-center/Tab";
import { MessageManager } from "../messages/MessageManager";
import { MessageDiv, MessageSpan } from "../messages/MessageSpan";
import type { NotifyMessageDetailsType } from "../messages/ReactNotifyMessageDetails";
import { StatusBar } from "../statusbar/StatusBar";

/** Enum for the [[MessageCenterField]] active tab
 * @internal
 */
enum MessageCenterActiveTab {
  AllMessages,
  Problems,
}

/** State for the [[MessageCenterField]] React component
 * @internal
 */
interface MessageCenterState {
  activeTab: MessageCenterActiveTab;
  target: HTMLDivElement | null;
  messageCount: number;
  isOpen: boolean;
}

/** Message Center Field React component.
 * @public
 */
export class MessageCenterField extends React.Component<
  CommonProps,
  MessageCenterState
> {
  private _indicator = React.createRef<HTMLDivElement>();
  private _title = UiFramework.translate("messageCenter.messages");
  private _unloadMessagesUpdatedHandler?: () => void;
  private _removeOpenMessagesCenterHandler?: () => void;

  constructor(p: CommonProps) {
    super(p);

    this.state = {
      activeTab: MessageCenterActiveTab.AllMessages,
      target: null,
      messageCount: MessageManager.messages.length,
      isOpen: false,
    };
  }

  /** @internal */
  public override componentDidMount() {
    this._unloadMessagesUpdatedHandler =
      MessageManager.onMessagesUpdatedEvent.addListener(
        this._handleMessagesUpdatedEvent,
        this
      );
    this._removeOpenMessagesCenterHandler =
      MessageManager.onOpenMessageCenterEvent.addListener(
        this._handleOpenMessageCenterEvent,
        this
      );
    MessageManager.registerAnimateOutToElement(this._indicator.current);
  }

  /** @internal */
  public override componentWillUnmount() {
    // istanbul ignore else
    if (this._unloadMessagesUpdatedHandler) {
      this._unloadMessagesUpdatedHandler();
      this._unloadMessagesUpdatedHandler = undefined;
    }
    // istanbul ignore else
    if (this._removeOpenMessagesCenterHandler) {
      this._removeOpenMessagesCenterHandler();
      this._removeOpenMessagesCenterHandler = undefined;
    }
  }

  private _handleMessagesUpdatedEvent = () => {
    // istanbul ignore else
    if (this._unloadMessagesUpdatedHandler)
      this.setState({ messageCount: MessageManager.messages.length });
  };

  private _handleOpenMessageCenterEvent = () => {
    this.setIsOpen(true);
  };

  public override render(): React.ReactNode {
    const tooltip = `${this.state.messageCount} ${this._title}`;
    const divStyle = { ...this.props.style, height: "100%" };
    const footerMessages = (
      <>
        <div className={this.props.className} style={divStyle} title={tooltip}>
          <MessageCenter
            indicatorRef={this._indicator}
            label={this._title}
            onClick={this._handleMessageIndicatorClick}
          >
            {this.state.messageCount.toString()}
          </MessageCenter>
        </div>
        <StatusBar.Popup
          showArrow={false}
          isOpen={this.state.isOpen}
          onClose={this._handleClose}
          onOutsideClick={this._handleOutsideClick}
          target={this._indicator.current}
        >
          <MessageCenterDialog
            prompt={UiFramework.translate("messageCenter.prompt")}
            // <>
            //   <MessageCenterTab
            //     isActive={
            //       this.state.activeTab === MessageCenterActiveTab.AllMessages
            //     }
            //     onClick={() =>
            //       this._changeActiveTab(MessageCenterActiveTab.AllMessages)
            //     }
            //   >
            //     {UiFramework.translate("messageCenter.all")}
            //   </MessageCenterTab>
            //   <MessageCenterTab
            //     isActive={
            //       this.state.activeTab === MessageCenterActiveTab.Problems
            //     }
            //     onClick={() =>
            //       this._changeActiveTab(MessageCenterActiveTab.Problems)
            //     }
            //   >
            //     {UiFramework.translate("messageCenter.errors")}
            //   </MessageCenterTab>
            // </>

            title={this._title}
          >
            <Tabs
              type="pill"
              labels={[
                <Tab key="All" label="All" />,
                <Tab key="Errors" label="Errors" />,
              ]}
            >
              {"content"}
            </Tabs>
            {this.getMessages()}
          </MessageCenterDialog>
        </StatusBar.Popup>
      </>
    );

    return footerMessages;
  }

  private _handleClose = () => {
    this.setIsOpen(false);
  };

  private _handleOutsideClick = (e: MouseEvent) => {
    if (
      !this._indicator.current ||
      !(e.target instanceof Node) ||
      this._indicator.current.contains(e.target)
    )
      return;

    this._handleClose();
  };

  private _handleMessageIndicatorClick = () => {
    this.setIsOpen(!this.state.isOpen);
  };

  private _changeActiveTab = (tab: MessageCenterActiveTab) => {
    this.setState({ activeTab: tab });
  };

  private getMessages(): React.ReactChild[] {
    const messages = MessageManager.messages.slice(0).reverse();
    const tabRows: React.ReactChild[] = new Array<React.ReactChild>();

    messages.forEach((details: NotifyMessageDetailsType, index: number) => {
      /* istanbul ignore else */
      if (
        this.state.activeTab === MessageCenterActiveTab.AllMessages ||
        this.isProblemStatus(details.priority)
      ) {
        const iconClassName = MessageManager.getIconClassName(details);
        const iconSpec = MessageManager.getIconSpecFromDetails(details);
        const message = details.briefMessage;

        tabRows.push(
          <MessageCenterMessage
            key={index.toString()}
            icon={<Icon iconSpec={iconSpec} className={iconClassName} />}
          >
            <MessageSpan message={message} />
            {details.detailedMessage && (
              <MessageDiv
                className="iui-text-small"
                message={details.detailedMessage}
              />
            )}
          </MessageCenterMessage>
        );
      }
    });

    return tabRows;
  }

  private isProblemStatus(priority: OutputMessagePriority): boolean {
    // See priority values in DgnPlatform defined in NotificationManager

    if (
      priority === OutputMessagePriority.Error ||
      priority === OutputMessagePriority.Fatal
    )
      return true;

    return false;
  }

  private setIsOpen(isOpen: boolean) {
    this.setState({ isOpen });
  }
}

/*


export const TEST : React.FC (props : <CommonProps,
  MessageCenterState>)=> {
  const {
    style,
    id,
    activeTab,
    target,
    messageCount,
    isOpen,
  } = props

  const _indicator = React.createRef<HTMLDivElement>();
  const _title = UiFramework.translate("messageCenter.messages");
  const _unloadMessagesUpdatedHandler?: () => void;
  const _removeOpenMessagesCenterHandler?: () => void;

  constructor(p: CommonProps) {
    super(p);

    this.state = {
      activeTab: MessageCenterActiveTab.AllMessages,
      target: null,
      messageCount: MessageManager.messages.length,
      isOpen: false,
    };
  }

  public override componentDidMount() {
    this._unloadMessagesUpdatedHandler =
      MessageManager.onMessagesUpdatedEvent.addListener(
        this._handleMessagesUpdatedEvent,
        this
      );
    this._removeOpenMessagesCenterHandler =
      MessageManager.onOpenMessageCenterEvent.addListener(
        this._handleOpenMessageCenterEvent,
        this
      );
    MessageManager.registerAnimateOutToElement(this._indicator.current);
  }


  public override componentWillUnmount() {
    // istanbul ignore else
    if (this._unloadMessagesUpdatedHandler) {
      this._unloadMessagesUpdatedHandler();
      this._unloadMessagesUpdatedHandler = undefined;
    }
    // istanbul ignore else
    if (this._removeOpenMessagesCenterHandler) {
      this._removeOpenMessagesCenterHandler();
      this._removeOpenMessagesCenterHandler = undefined;
    }
  }

  private _handleMessagesUpdatedEvent = () => {
    // istanbul ignore else
    if (this._unloadMessagesUpdatedHandler)
      this.setState({ messageCount: MessageManager.messages.length });
  };

  private _handleOpenMessageCenterEvent = () => {
    this.setIsOpen(true);
  };

  public override render(): React.ReactNode {
    const tooltip = `${this.state.messageCount} ${this._title}`;
    const divStyle = { ...this.props.style, height: "100%" };
    const footerMessages = (
      <>
        <div className={this.props.className} style={divStyle} title={tooltip}>
          <MessageCenter
            indicatorRef={this._indicator}
            label={this._title}
            onClick={this._handleMessageIndicatorClick}
          >
            {this.state.messageCount.toString()}
          </MessageCenter>
        </div>
        <StatusBar.Popup
          isOpen={this.state.isOpen}
          onClose={this._handleClose}
          onOutsideClick={this._handleOutsideClick}
          target={this._indicator.current}
        >
          <MessageCenterDialog
            prompt={UiFramework.translate("messageCenter.prompt")}
            tabs={
              <>
                <MessageCenterTab
                  isActive={
                    this.state.activeTab === MessageCenterActiveTab.AllMessages
                  }
                  onClick={() =>
                    this._changeActiveTab(MessageCenterActiveTab.AllMessages)
                  }
                >
                  {UiFramework.translate("messageCenter.all")}
                </MessageCenterTab>
                <MessageCenterTab
                  isActive={
                    this.state.activeTab === MessageCenterActiveTab.Problems
                  }
                  onClick={() =>
                    this._changeActiveTab(MessageCenterActiveTab.Problems)
                  }
                >
                  {UiFramework.translate("messageCenter.errors")}
                </MessageCenterTab>
              </>
            }
            title={this._title}
          >
            {this.getMessages()}
          </MessageCenterDialog>
        </StatusBar.Popup>
      </>
    );

    return footerMessages;
  }

  private _handleClose = () => {
    this.setIsOpen(false);
  };

  private _handleOutsideClick = (e: MouseEvent) => {
    if (
      !this._indicator.current ||
      !(e.target instanceof Node) ||
      this._indicator.current.contains(e.target)
    )
      return;

    this._handleClose();
  };

  private _handleMessageIndicatorClick = () => {
    this.setIsOpen(!this.state.isOpen);
  };

  private _changeActiveTab = (tab: MessageCenterActiveTab) => {
    this.setState({ activeTab: tab });
  };

  private getMessages(): React.ReactChild[] {
    const messages = MessageManager.messages.slice(0).reverse();
    const tabRows: React.ReactChild[] = new Array<React.ReactChild>();

    messages.forEach((details: NotifyMessageDetailsType, index: number) => {

      if (
        this.state.activeTab === MessageCenterActiveTab.AllMessages ||
        this.isProblemStatus(details.priority)
      ) {
        const iconClassName = MessageManager.getIconClassName(details);
        const iconSpec = MessageManager.getIconSpecFromDetails(details);
        const message = details.briefMessage;

        tabRows.push(
          <MessageCenterMessage
            key={index.toString()}
            icon={<Icon iconSpec={iconSpec} className={iconClassName} />}
          >
            <MessageSpan message={message} />
            {details.detailedMessage && (
              <MessageDiv
                className="iui-text-small"
                message={details.detailedMessage}
              />
            )}
          </MessageCenterMessage>
        );
      }
    });

    return tabRows;
  }

  private isProblemStatus(priority: OutputMessagePriority): boolean {
    // See priority values in DgnPlatform defined in NotificationManager

    if (
      priority === OutputMessagePriority.Error ||
      priority === OutputMessagePriority.Fatal
    )
      return true;

    return false;
  }

  private setIsOpen(isOpen: boolean) {
    this.setState({ isOpen });
  }
}


*/
