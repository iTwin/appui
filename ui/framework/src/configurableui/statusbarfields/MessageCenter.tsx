/*---------------------------------------------------------------------------------------------
| $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
/** @module Notification */

import * as React from "react";

import { NotifyMessageDetails, OutputMessagePriority } from "@bentley/imodeljs-frontend";

import "../configurableui.scss";

import UiFramework from "../../UiFramework";

import { StatusBarFieldId, IStatusBar } from "../StatusBarWidgetControl";
import { MessageManager } from "../MessageManager";

import MessageCenterIndicator from "@bentley/ui-ninezone/lib/footer/message-center/Indicator";
import MessageCenterTab from "@bentley/ui-ninezone/lib/footer/message-center/Tab";
import MessageCenterMessage from "@bentley/ui-ninezone/lib/footer/message-center/Message";
import MessageCenter, { MessageCenterButton } from "@bentley/ui-ninezone/lib/footer/message-center/MessageCenter";

export interface MessageCenterProps {
  statusBar: IStatusBar;
  isInFooterMode: boolean;
  openWidget: StatusBarFieldId;
}

export enum MessageCenterActiveTab {
  AllMessages,
  Problems,
}

export interface MessageCenterState {
  activeTab: MessageCenterActiveTab;
}

/** Message Center Field React component.
Â */
export class MessageCenterField extends React.Component<MessageCenterProps, MessageCenterState> {
  private _className: string;
  private _element: any;

  public readonly state: Readonly<MessageCenterState> = {
    activeTab: MessageCenterActiveTab.AllMessages,
  };

  constructor(p: MessageCenterProps) {
    super(p);

    const instance = this.constructor;
    this._className = instance.name;
  }

  public render(): React.ReactNode {
    const messageCount = MessageManager.Messages.length;
    const footerMessages = (
      <MessageCenterIndicator
        ref={(element: any) => { this._element = element; }}
        label={UiFramework.i18n.translate("UiFramework:messageCenter.messages")}
        isLabelVisible={this.props.isInFooterMode}
        balloonText={messageCount.toString()}
        onIndicatorClick={this.handleMessageIndicatorClick}
        dialog={
          <MessageCenter
            isOpen={this.props.openWidget === this._className}
            title={UiFramework.i18n.translate("UiFramework:messageCenter.messages")}
            buttons={
              <>
                <MessageCenterButton>
                  <i className={"icon icon-export"} />
                </MessageCenterButton>
                <MessageCenterButton onClick={this.handleCloseMessageIndicatorClick}>
                  <i className={"icon icon-close"} />
                </MessageCenterButton>
              </>
            }
            tabs={
              <>
                <MessageCenterTab
                  isOpen={this.state.activeTab === MessageCenterActiveTab.AllMessages}
                  onClick={() => this.changeActiveTab(MessageCenterActiveTab.AllMessages)}
                >
                  {UiFramework.i18n.translate("UiFramework:messageCenter.all")}
                </MessageCenterTab>
                <MessageCenterTab
                  isOpen={this.state.activeTab === MessageCenterActiveTab.Problems}
                  onClick={() => this.changeActiveTab(MessageCenterActiveTab.Problems)}
                >
                  {UiFramework.i18n.translate("UiFramework:messageCenter.problems")}
                </MessageCenterTab>
              </>
            }
            messages={this.getMessages()}
          />
        }
      />
    );

    this.props.statusBar.setFooterMessages(this._element);

    return footerMessages;
  }

  private handleMessageIndicatorClick = () => {
    const isOpen = this.props.openWidget === this._className;
    if (isOpen)
      this.setOpenWidget(null);
    else
      this.setOpenWidget(this._className);
  }

  private handleCloseMessageIndicatorClick = () => {
    this.setOpenWidget(null);
  }

  private changeActiveTab = (tab: MessageCenterActiveTab) => {
    this.setState((_prevState) => {
      return {
        activeTab: tab,
      };
    });
  }

  private getMessages(): React.ReactChild[] {
    const messages = MessageManager.Messages.slice(0).reverse();
    const tabRows: React.ReactChild[] = new Array<React.ReactChild>();

    messages.forEach((details: NotifyMessageDetails, index: number) => {
      if (this.state.activeTab === MessageCenterActiveTab.AllMessages || this.isProblemStatus(details.priority)) {

        const iconClassName = MessageManager.getIconClassName(details);

        tabRows.push(
          <MessageCenterMessage
            key={index.toString()}
            icon={<i className={iconClassName} />}
            content={<span>{details.briefMessage}</span>}
          />,
        );
      }
    });

    return tabRows;
  }

  private isProblemStatus(priority: OutputMessagePriority): boolean {
    // See priority values in DgnPlatform defined in NotificationManager

    if (priority === OutputMessagePriority.Error || priority === OutputMessagePriority.Fatal)
      return true;

    return false;
  }

  private setOpenWidget(openWidget: StatusBarFieldId) {
    this.props.statusBar.setOpenWidget(openWidget);
  }

}
