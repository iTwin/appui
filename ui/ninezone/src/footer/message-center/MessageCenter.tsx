/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module MessageCenter */

import * as classnames from "classnames";
import * as React from "react";
import withContainInViewport from "../../base/WithContainInViewport";
import CommonProps, { NoChildrenProps } from "../../utilities/Props";
import Direction from "../../utilities/Direction";
import Popover from "../../popup/popover/Triangle";
import Dialog from "../message/content/dialog/Dialog";
import TitleBar from "../message/content/dialog/TitleBar";
import Title from "../message/content/dialog/Title";
export { DialogButton as MessageCenterButton } from "../message/content/dialog/Button";
import Content from "./Content";
import "./MessageCenter.scss";

// tslint:disable-next-line:variable-name
const DialogWithContainIn = withContainInViewport(Dialog);

/** Properties of [[MessageCenter]] component. */
export interface MessageCenterProps extends CommonProps, NoChildrenProps {
  /** Title of title bar. */
  title?: string;
  /** Title bar buttons. I.e.: [[DialogButton]] */
  buttons?: React.ReactNode;
  /** Tabs of message center. See [[MessageCenterTab]] */
  tabs?: React.ReactNode;
  /** Messages of message center. I.e. [[Message]] */
  messages?: React.ReactNode;
  /* Optional prompt when no messages are present */
  prompt?: string;
}

/** Message center dialog used in [[MessageCenterIndicator]] component. */
// tslint:disable-next-line:variable-name
export const MessageCenter: React.StatelessComponent<MessageCenterProps> = (props) => {
  const dialogClassName = classnames(
    "nz-footer-messageCenter-messageCenter",
    props.className);

  return (
    <Popover
      className={dialogClassName}
      style={props.style}
      direction={Direction.Top}
      content={
        <DialogWithContainIn
          noVerticalContainment
          titleBar={
            <TitleBar
              title={
                <Title text={props.title} />
              }
              buttons={props.buttons}
            />
          }
          content={
          <Content
            tabs={props.tabs}
            messages={props.messages}
            prompt={props.prompt}
            />
          }
        />
      }
    />
  );
};

export default MessageCenter;
