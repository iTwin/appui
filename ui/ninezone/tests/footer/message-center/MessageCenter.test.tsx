/*---------------------------------------------------------------------------------------------
|  $Copyright: (c) 2018 Bentley Systems, Incorporated. All rights reserved. $
 *--------------------------------------------------------------------------------------------*/
import { mount, shallow } from "enzyme";
import * as React from "react";

import MessageCenter from "../../../src/footer/message-center/MessageCenter";

describe("<MessageCenter />", () => {
  it("should render", () => {
    mount(<MessageCenter />);
  });

  it("renders correctly", () => {
    shallow(<MessageCenter />).should.matchSnapshot();
  });
});
