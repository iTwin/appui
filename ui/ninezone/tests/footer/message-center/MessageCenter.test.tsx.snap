// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageCenter /> renders correctly 1`] = `
<Component
  className="nz-footer-messageCenter-messageCenter"
  content={
    <WithContainInViewport
      content={
        <Unknown
          messages={undefined}
          prompt={undefined}
          tabs={undefined}
        />
      }
      noVerticalContainment={true}
      titleBar={
        <Unknown
          buttons={undefined}
          title={
            <Unknown
              text={undefined}
            />
          }
        />
      }
    />
  }
  direction={1}
/>
`;
